### PR TITLE
[R4R] Add cdp simulations

### PR DESCRIPTION
**Pull request by karzak**

### CODE REVIEW

File: app/sim_test.go [hunk 1]
```diff
@@ -304,6 +306,17 @@ func testAndRunTxs(app *App, config simulation.Config) []simulation.WeightedOper
 			}(nil),
 			pricefeedsimops.SimulateMsgUpdatePrices(app.pricefeedKeeper),
 		},
+		{
+			func(_ *rand.Rand) int {
+				var v int
+				ap.GetOrGenerate(app.cdc, OpWeightMsgCdp, &v, nil,
+					func(_ *rand.Rand) {
+						v = 10000 // TODO
```
Comment (fedekunze) [inline app/sim_test.go:314]: this is the percentage of simulation (i.e  from all the times this operation is selected, this is the % of the times this message is actually run)

Comment (fedekunze) [reply app/sim_test.go:314]:
```suggestion
						v = 100
```

File: x/cdp/simulation/genesis.go [hunk 2]
```diff
@@ -4,19 +4,164 @@ import (
...
+	for i, a := range accounts {
+		if a.GetAddress().Equals(acc.GetAddress()) {
+			newAccounts[i] = acc
+			return newAccounts
+		}
+	}
+	return append(newAccounts, acc)
+}
+
+func randomCdpGenState(selection int) types.GenesisState {
```
Comment (fedekunze) [inline x/cdp/simulation/genesis.go:66]: I guess is ok for now to have the genesis state hardcoded. In the future, we might want to make it actually random with different collateral params, thresholds, etc.

File: x/cdp/simulation/operations/msgs.go [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+)
+
+// SimulateMsgCdp generates a MsgCreateCdp or MsgDepositCdp with random values.
+func SimulateMsgCdp(ak auth.AccountKeeper, k cdp.Keeper, pfk pricefeed.Keeper) simulation.Operation {
+	return func(r *rand.Rand, app *baseapp.BaseApp, ctx sdk.Context, accs []simulation.Account) (
+		opMsg simulation.OperationMsg, fOps []simulation.FutureOperation, err error) {
+
+		handler := cdp.NewHandler(k)
+		simacc := simulation.RandomAcc(r, accs)
+		acc := ak.GetAccount(ctx, simacc.Address)
```
Comment (fedekunze) [inline x/cdp/simulation/operations/msgs.go:23]: check for `acc == nil`

File: x/cdp/simulation/operations/msgs.go [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+		randDebtAsset := randCollateralParam.DebtLimit[r.Intn(len(randCollateralParam.DebtLimit))]
+		randDebtParam, _ := k.GetDebtParam(ctx, randDebtAsset.Denom)
+		if coins.AmountOf(randCollateralParam.Denom).IsZero() {
+			return simulation.NoOpMsg(cdp.ModuleName), nil, nil
+		}
+
+		price, err := pfk.GetCurrentPrice(ctx, randCollateralParam.MarketID)
+		if err != nil {
+			fmt.Printf("error from pricefeed: %v\n", err)
+			return simulation.NoOpMsg(cdp.ModuleName), nil, nil
```
Comment (fedekunze) [inline x/cdp/simulation/operations/msgs.go:39]:
```suggestion
			return simulation.NoOpMsg(cdp.ModuleName), nil, err
```

File: x/cdp/simulation/operations/msgs.go [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+	}
+}
+
+func submitMsg(msg sdk.Msg, handler sdk.Handler, ctx sdk.Context) (ok bool) {
+	ctx, write := ctx.CacheContext()
+	ok = handler(ctx, msg).IsOK()
+	if ok {
+		write()
+	} else {
+		log := handler(ctx, msg).Log
```
Comment (fedekunze) [inline x/cdp/simulation/operations/msgs.go:167]: this will run the msg again
```suggestion
```

File: x/cdp/simulation/operations/msgs.go [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+			return simulation.NewOperationMsg(msg, ok, "repay debt cdp"), nil, nil
+		}
+
+		return simulation.NewOperationMsgBasic(cdp.ModuleName, "no-operation (no valid actions)", "", false, nil), nil, nil
+	}
+}
+
+func submitMsg(msg sdk.Msg, handler sdk.Handler, ctx sdk.Context) (ok bool) {
+	ctx, write := ctx.CacheContext()
+	ok = handler(ctx, msg).IsOK()
```
Comment (fedekunze) [inline x/cdp/simulation/operations/msgs.go:163]:
```suggestion
        res := handler(ctx, msg)
```

File: x/cdp/simulation/operations/msgs.go [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+	if ok {
```
Comment (fedekunze) [inline x/cdp/simulation/operations/msgs.go:164]:
```suggestion
	if res.IsOK() {
```

File: x/cdp/simulation/operations/msgs.go [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+		write()
+	} else {
+		log := handler(ctx, msg).Log
+		fmt.Println(log)
```
Comment (fedekunze) [inline x/cdp/simulation/operations/msgs.go:168]:
```suggestion
		fmt.Println(res.Log)
```

File: x/cdp/simulation/operations/msgs.go [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+	}
+	return ok
```
Comment (fedekunze) [inline x/cdp/simulation/operations/msgs.go:170]:
```suggestion
	return res.IsOK()
```

File: app/app.go [hunk 10]
```diff
@@ -329,9 +329,9 @@ func NewApp(logger log.Logger, db dbm.DB, traceStore io.Writer, loadLatest bool,
 		mint.NewAppModule(app.mintKeeper),
 		distr.NewAppModule(app.distrKeeper, app.supplyKeeper),
 		staking.NewAppModule(app.stakingKeeper, app.accountKeeper, app.supplyKeeper),
-		slashing.NewAppModule(app.slashingKeeper, app.stakingKeeper),
-		cdp.NewAppModule(app.cdpKeeper, app.pricefeedKeeper, app.supplyKeeper), // TODO how is the order be decided here? Is this order correct?
+		slashing.NewAppModule(app.slashingKeeper, app.stakingKeeper), // TODO how is the order be decided here? Is this order correct?
```
Comment (fedekunze) [inline app/app.go:332]:
```suggestion
		slashing.NewAppModule(app.slashingKeeper, app.stakingKeeper),
```

File: x/cdp/simulation/operations/msgs.go [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+	value := simulation.RandIntBetween(r, 1, 100)
+	if value > threshold {
+		return true
+	}
+	return false
+}
+
+func canClose(acc authexported.Account, c cdp.CDP, denom string) bool {
+	repaymentAmount := c.Principal.Add(c.AccumulatedFees).AmountOf(denom)
+	if acc.GetCoins().AmountOf(denom).GT(repaymentAmount) {
```
Comment (denalimarsh) [inline x/cdp/simulation/operations/msgs.go:210]:
```suggestion
	if acc.GetCoins().AmountOf(denom).GTE(repaymentAmount) {
```

File: x/cdp/simulation/operations/msgs.go [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+				return simulation.NoOpMsg(cdp.ModuleName), nil, fmt.Errorf("expected msg to pass ValidateBasic: %s", msg.GetSignBytes())
+			}
+			ok := submitMsg(msg, handler, ctx)
+			if !ok {
+				return simulation.NoOpMsg(cdp.ModuleName), nil, fmt.Errorf("could not submit create cdp msg")
+			}
+			return simulation.NewOperationMsg(msg, ok, "create cdp"), nil, nil
+		}
+
+		// TODO a cdp already exists, deposit to it, draw debt from it, or repay debt to it
```
Comment (denalimarsh) [inline x/cdp/simulation/operations/msgs.go:81]:
```suggestion
		a cdp already exists, deposit to it, draw debt from it, or repay debt to it
```

File: x/cdp/simulation/operations/msgs.go [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+			randRepayAmount := sdk.NewInt(int64(simulation.RandIntBetween(r, 1, int(maxRepay.Int64()))))
+			if debt.Equal(randDebtParam.DebtFloor) {
+				if acc.GetCoins().AmountOf(randDebtParam.Denom).GTE(debt) {
+					randRepayAmount = debt
+				}
+			}
+			msg := cdp.NewMsgRepayDebt(acc.GetAddress(), randCollateralParam.Denom, sdk.NewCoins(sdk.NewCoin(randDebtParam.Denom, randRepayAmount)))
+			err := msg.ValidateBasic()
+			if err != nil {
+				return simulation.NoOpMsg(cdp.ModuleName), nil, fmt.Errorf("expected repay msg to pass ValidateBasic: %s", msg.GetSignBytes())
```
Comment (denalimarsh) [inline x/cdp/simulation/operations/msgs.go:150]:
```suggestion
				return simulation.NoOpMsg(cdp.ModuleName), nil, fmt.Errorf("expected repay msg to pass ValidateBasic: %v", err)
```

File: x/cdp/simulation/operations/msgs.go [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+			collateralDeposit := sdk.NewInt(int64(simulation.RandIntBetween(r, int(minCollateralDepositRounded.Int64()), int(maxCollateralDeposit.Int64()))))
+			// calculate how much the randomly selected deposit is worth
+			collateralDepositValue := ShiftDec(sdk.NewDecFromInt(collateralDeposit), randCollateralParam.ConversionFactor.Neg()).Mul(priceShifted)
+			// calculate the max amount of debt that could be drawn for the chosen deposit
+			maxDebtDraw := collateralDepositValue.Quo(randCollateralParam.LiquidationRatio).TruncateInt()
+			// randomly select a debt draw amount
+			debtDraw := sdk.NewInt(int64(simulation.RandIntBetween(r, int(randDebtParam.DebtFloor.Int64()), int(maxDebtDraw.Int64()))))
+			msg := cdp.NewMsgCreateCDP(acc.GetAddress(), sdk.NewCoins(sdk.NewCoin(randCollateralParam.Denom, collateralDeposit)), sdk.NewCoins(sdk.NewCoin(randDebtParam.Denom, debtDraw)))
+			if msg.ValidateBasic() != nil {
+				return simulation.NoOpMsg(cdp.ModuleName), nil, fmt.Errorf("expected msg to pass ValidateBasic: %s", msg.GetSignBytes())
```
Comment (denalimarsh) [inline x/cdp/simulation/operations/msgs.go:72]:
```suggestion
				return simulation.NoOpMsg(cdp.ModuleName), nil, fmt.Errorf("expected create msg to pass ValidateBasic: %v", err)
```

File: x/cdp/simulation/operations/msgs.go [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+			// randomly select a collateral deposit amount
+			collateralDeposit := sdk.NewInt(int64(simulation.RandIntBetween(r, int(minCollateralDepositRounded.Int64()), int(maxCollateralDeposit.Int64()))))
+			// calculate how much the randomly selected deposit is worth
+			collateralDepositValue := ShiftDec(sdk.NewDecFromInt(collateralDeposit), randCollateralParam.ConversionFactor.Neg()).Mul(priceShifted)
+			// calculate the max amount of debt that could be drawn for the chosen deposit
+			maxDebtDraw := collateralDepositValue.Quo(randCollateralParam.LiquidationRatio).TruncateInt()
+			// randomly select a debt draw amount
+			debtDraw := sdk.NewInt(int64(simulation.RandIntBetween(r, int(randDebtParam.DebtFloor.Int64()), int(maxDebtDraw.Int64()))))
+			msg := cdp.NewMsgCreateCDP(acc.GetAddress(), sdk.NewCoins(sdk.NewCoin(randCollateralParam.Denom, collateralDeposit)), sdk.NewCoins(sdk.NewCoin(randDebtParam.Denom, debtDraw)))
+			if msg.ValidateBasic() != nil {
```
Comment (denalimarsh) [inline x/cdp/simulation/operations/msgs.go:71]:
```suggestion
			err := msg.ValidateBasic()
			if err != nil {
```

**Pull request merged**